### PR TITLE
add zdb support

### PR DESCRIPTION
**Pull request by qwdarrenyin**
add zdb support

### CODE REVIEW

File: libledger/DBInitializer.cpp [hunk 1]
```diff
@@ -246,6 +253,25 @@ void DBInitializer::initRocksDBStorage()
...
+    DBInitializer_LOG(INFO) << LOG_BADGE("initStorageDB") << LOG_BADGE("initZdbStorage");
+    auto zdbStorage = std::make_shared<ZdbStorage>();
+
+    zdbStorage->initSqlAccess(m_param->mutableStorageParam().dbtype,
+        m_param->mutableStorageParam().dbip, m_param->mutableStorageParam().dbport,
+        m_param->mutableStorageParam().dbusername, m_param->mutableStorageParam().dbpasswd,
+        m_param->mutableStorageParam().dbname, m_param->mutableStorageParam().dbcharset,
+        m_param->mutableStorageParam().initconnections,
+        m_param->mutableStorageParam().maxconnections);
+    /*m_storage = zdbStorage;
```
Comment (bxq2011hust) [inline libledger/DBInitializer.cpp]: delete this

File: libledger/Ledger.cpp [hunk 2]
```diff
@@ -329,11 +329,11 @@ void Ledger::initDBConfig(ptree const& pt)
 {
     /// init the basic config
     /// set storage db related param
-    m_param->mutableStorageParam().type = pt.get<std::string>("storage.type", "RocksDB");
+    m_param->mutableStorageParam().type = pt.get<std::string>("storage.type", "zdbstorage");
```
Comment (bxq2011hust) [inline libledger/Ledger.cpp]: default should be RocksDB

File: libledger/Ledger.cpp [hunk 3]
```diff
@@ -351,11 +351,33 @@ void Ledger::initDBConfig(ptree const& pt)
     {
         m_param->mutableStorageParam().maxRetry = 100;
     }
+    /// set state db related param
+    m_param->mutableStateParam().type = pt.get<std::string>("state.type", "storage");
+
+    // read db config from config eg:mysqlip mysqlport and so on
+    m_param->mutableStorageParam().dbtype = pt.get<std::string>("storage.dbtype", "mysql");
```
Comment (bxq2011hust) [inline libledger/Ledger.cpp]: this config item is unnecessary, we only support mysql

File: libledger/Ledger.cpp [hunk 4]
```diff
@@ -351,11 +351,33 @@ void Ledger::initDBConfig(ptree const& pt)
...
     }
+    /// set state db related param
+    m_param->mutableStateParam().type = pt.get<std::string>("state.type", "storage");
+
+    // read db config from config eg:mysqlip mysqlport and so on
+    m_param->mutableStorageParam().dbtype = pt.get<std::string>("storage.dbtype", "mysql");
+    m_param->mutableStorageParam().dbip = pt.get<std::string>("storage.dbip", "127.0.0.1");
+    m_param->mutableStorageParam().dbport = pt.get<int>("storage.dbport", 3306);
+    m_param->mutableStorageParam().dbusername = pt.get<std::string>("storage.dbusername", "root");
+    m_param->mutableStorageParam().dbpasswd = pt.get<std::string>("storage.dbpasswd", "123456");
```
Comment (bxq2011hust) [inline libledger/Ledger.cpp]: delete passwd

File: libledger/Ledger.cpp (hunk 3 above)
Comment (qwdarrenyin) [reply libledger/Ledger.cpp]: actually we can support sqllite,oracle and postgresql if necessary.

File: libledger/Ledger.cpp (hunk 2 above)
Comment (qwdarrenyin) [reply libledger/Ledger.cpp]: get

File: libledger/DBInitializer.cpp (hunk 1 above)
Comment (qwdarrenyin) [reply libledger/DBInitializer.cpp]: get

File: cmake/ProjectLibzdb.cmake [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    CONFIGURE_COMMAND ${configure_command} 
+    BUILD_COMMAND make
+    INSTALL_COMMAND ""
+    BUILD_BYPRODUCTS <SOURCE_DIR>/.libs/libzdb.a
+)
+
+ExternalProject_Get_Property(libzdb SOURCE_DIR)
+
+add_library(zdb STATIC IMPORTED GLOBAL)
+set(LIBZDB_INCLUDE_DIR ${SOURCE_DIR}/zdb/)
```
Comment (bxq2011hust) [inline cmake/ProjectLibzdb.cmake]: delete zdb

File: libdevcore/SQLConnectionPool.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ */
+/** @file SQLConnectionPool.h
+ *  @author darrenyin
+ *  @date 2019-04-24
+ */
+
+
+#pragma once
+
+#include <Connection.h>
```
Comment (bxq2011hust) [inline libdevcore/SQLConnectionPool.h]: modify to “zdb/Connection.h”, more specific

File: libledger/Ledger.cpp (hunk 3 above)
Comment (bxq2011hust) [reply libledger/Ledger.cpp]: I know that, but we will not to do that. If we do, we can add this configuration

File: libdevcore/SQLConnectionPool.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    const std::string& strDbname, const std::string& strDbCharset, uint32_t dwInitConnections,
+    uint32_t dwMaxConnection)
+{
+    if (strDbType == "mysql")
+    {
+        char cConnectionBuf[2048] = {0};
+        snprintf(cConnectionBuf, sizeof(cConnectionBuf),
+            "mysql://%s:%u/%s?user=%s&password=%s&charset=%s", strDbIp.c_str(), dwDbPort,
+            strDbname.c_str(), strDbUser.c_str(), strDbPasswd.c_str(), strDbCharset.c_str());
+        string strConnetionBuf = cConnectionBuf;
```
Comment (wheatli) [inline libdevcore/SQLConnectionPool.cpp]: why does it need to transfer from char* to string and reverse again ?

File: libledger/LedgerParam.h [hunk 8]
```diff
@@ -84,7 +84,18 @@ struct StorageParam
     // for amop storage
     std::string topic;
     size_t timeout;
-    int maxRetry;
+    size_t maxRetry;
```
Comment (wheatli) [inline libledger/LedgerParam.h]: standard code style is max_retry, db_type, and so on...

File: libledger/LedgerParam.h [hunk 9]
```diff
@@ -84,7 +84,18 @@ struct StorageParam
...
+
+    // for zdb storage
+    std::string dbtype;
+    std::string dbip;
+    uint32_t dbport;
+    std::string dbusername;
+    std::string dbpasswd;
+    std::string dbname;
+    std::string dbcharset;
```
Comment (wheatli) [inline libledger/LedgerParam.h]: same as last one

File: libledger/Ledger.cpp [hunk 10]
```diff
@@ -351,11 +351,33 @@ void Ledger::initDBConfig(ptree const& pt)
...
+    m_param->mutableStateParam().type = pt.get<std::string>("state.type", "storage");
+
+    // read db config from config eg:mysqlip mysqlport and so on
+    m_param->mutableStorageParam().dbtype = pt.get<std::string>("storage.dbtype", "mysql");
+    m_param->mutableStorageParam().dbip = pt.get<std::string>("storage.dbip", "127.0.0.1");
+    m_param->mutableStorageParam().dbport = pt.get<int>("storage.dbport", 3306);
+    m_param->mutableStorageParam().dbusername = pt.get<std::string>("storage.dbusername", "");
+    m_param->mutableStorageParam().dbpasswd = pt.get<std::string>("storage.dbpasswd", "");
+    m_param->mutableStorageParam().dbname = pt.get<std::string>("storage.dbname", "");
+    m_param->mutableStorageParam().dbcharset = pt.get<std::string>("storage.dbcharset", "utf8");
```
Comment (bxq2011hust) [inline libledger/Ledger.cpp]: https://stackoverflow.com/questions/30074492/what-is-the-difference-between-utf8mb4-and-utf8-charsets-in-mysql
is this ok to set utf8?

File: libstorage/ZdbStorage.cpp [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            -1, "Remote database return error:" + boost::lexical_cast<std::string>(iRet));
+    }
+
+
+    LOG(DEBUG) << "select resp:" << responseJson.toStyledString();
+    std::vector<std::string> columns;
+    for (Json::ArrayIndex i = 0; i < responseJson["result"]["columns"].size(); ++i)
+    {
+        std::string fieldName = responseJson["result"]["columns"].get(i, "").asString();
+        columns.push_back(fieldName);
```
Comment (wheatli) [inline libstorage/ZdbStorage.cpp]: too many copy here
1,asString 2, copy to fieldName 3,copy to columns

File: libdevcore/SQLConnectionPool.cpp (hunk 7 above)
Comment (bxq2011hust) [reply libdevcore/SQLConnectionPool.cpp]: changed.

File: libledger/LedgerParam.h (hunk 9 above)
Comment (bxq2011hust) [reply libledger/LedgerParam.h]: modified

File: libstorage/ZdbStorage.cpp (hunk 11 above)
Comment (bxq2011hust) [reply libstorage/ZdbStorage.cpp]: fixed

**Pull request merged**